service/storage_service.cc

```diff
@@ -317,6 +317,20 @@ future<> storage_service::topology_state_load(cdc::generation_service& cdc_gen_s
         co_return *ip;
     };
 
+    // We must check here that the node supports all features enabled
+    // in the cluster so far. If we won't do that then there is a risk
+    // that further commands might depend on some features that the current
+    // node doesn't understand and it might risk misinterpreting them.
+    //
+    // An interesting and useful side effect of those checks is that
+    // performing a successful read barrier makes sure that the local node
+    // supports all cluster's features.
```

Comment:
I would appreciate refactoring of the topology_state_load() and splitting it to load() and reload() where reload() automatically detects the delta in the topology change and only invokes state change triggers if the delta is present. It will make much easier to reason that various state change triggers have reasonable complexity, e.g. in this specific case it would not matter whether or not enable() is a no-op when features don't change, because it won't be invoked if features don't change. Not part of this patch scope.